src/containers/Application.jsx

```diff
@@ -142,22 +147,83 @@ class Application extends Component {
-        const provider = this.state.web3 ? this.state.web3.currentProvider : undefined;
-        return GivethWallet.loadWallet(keystore, provider);
-      })
-      .then(wallet => {
-        getWeb3().then(web3 => web3.setWallet(wallet));
-        this.setState({ wallet });
+      const { web3 } = window;
+
+      // check if web3 has been injected into dom window
+      if (typeof web3 !== 'undefined') {
+        const web3js = new Web3(web3.currentProvider);
+  
+        web3js.eth.getAccounts((err, acc) => {
+          const wallet = new BaseWallet(acc[0], web3js);
+  
+          // add wallet to application state
+          getWeb3().then(web3 => web3.setWallet(wallet));
+          this.setState({ wallet });
+        })
+    } else {
+      // if not web3 is not injected, use same Giveth wallet as before
+      GivethWallet.getCachedKeystore()
+        .then(keystore => {
+          // TODO change to getWeb3() when implemented. actually remove provider from GivethWallet
+          const provider = this.state.web3 ? this.state.web3.currentProvider : undefined;
+          return GivethWallet.loadWallet(keystore, provider);
+        })
+        .then(wallet => {
+          getWeb3().then(web3 => web3.setWallet(wallet));
+          this.setState({ wallet });
+        })
+        .catch(err => {
+          if (err.message !== 'No keystore found') console.error(err); // eslint-disable-line no-console
+        });
+    }
+  }
+
+  authMetaMask() {
+    const { web3 } = window;
+
+    // check if web3 has been injected into dom window
+    if (typeof web3 !== 'undefined') {
+      const web3js = new Web3(web3.currentProvider);
+
+      web3js.eth.getAccounts((err, acc) => {
+        // don't authenticate if there's already a user or if metamask is unlocked
+        if (this.state.currentUser) return; // fix this
+        if (!acc[0]) {
+          console.log('Show message about unlocking metamask')
+          return;
+        };
+
+        const authData = {
+          strategy: 'web3',
+          address: acc[0],
+        };
+        const util = new Accounts();
+        // console.log(test.hashMessage('uRr1tofqbXXG4yAEa2ed'))
+        return new Promise((resolve, reject) => {
+          feathersClient.authenticate(authData).catch(response => {
+            // normal flow will issue a 401 with a challenge message we need to sign and send to
+            // verify our identity
+            if (response.code === 401 && response.data.startsWith('Challenge =')) {
+              const msg = response.data.replace('Challenge =', '').trim();
+              resolve(web3js.eth.sign(util.hashMessage(msg), acc[0]));
+            }
+            reject(response);
+          });
+        })
+          .then(signature => {
+            authData.signature = signature;
+            return feathersClient.authenticate(authData);
+          })
+          .then(response => {
+            this.onSignIn(acc[0]);
+          })
+          .catch(err => console.log(err))
```

Comment:
all of this metamask code should be moved to the `InjectedWallet` implementation.

Instead of `GivethWallet.getCachedKeystore()` we can have a method on the interface `Wallet.load` or something similar where all this logic takes place.